### PR TITLE
Clear Vocab Sheets that are more than one month old.

### PR DESCRIPTION
**Pull request by KatieWright26**
[See zendesk ticked here](https://rabidtech.zendesk.com/agent/tickets/205)

Added dependent to VocabSheet's has_many relationship to ensure that associated items are also removed if a vocab sheet is deleted.

### CODE REVIEW

File: lib/tasks/clear_vocab_sheets.rake [hunk 1]
```diff
@@ -0,0 +1,12 @@
+namespace :vocab_sheets do
+  MAX_DAYS = 31
+
+  desc 'Clears all vocab sheets that are more than one month old'
+  task clear: :environment do
+    total_destroyed = 0
+    total_destroyed += VocabSheet.where('created_at < ?', MAX_DAYS.days.ago).destroy_all.length
```
Comment (KatieWright26) [inline lib/tasks/clear_vocab_sheets.rake]: Might be smarter to change this to 'updated_at < ?' ?

Comment (Br3nda) [reply lib/tasks/clear_vocab_sheets.rake]: is this was a class method on the VocabSheet model instead, you could write some nice tests.

File: app/models/vocab_sheet.rb [hunk 2]
```diff
@@ -1,7 +1,7 @@
 ##
 # A "sheet" of items (signs) saved by a user
 class VocabSheet < ActiveRecord::Base
-  has_many :items
+  has_many :items, dependent: :destroy
```
Comment (ben-rabid) [inline app/models/vocab_sheet.rb]: :+1:

File: lib/tasks/clear_vocab_sheets.rake [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+
+    puts "A total of #{total_destroyed} vocabulary sheets were deleted, #{VocabSheet.count} remaining"\
```
Comment (ben-rabid) [inline lib/tasks/clear_vocab_sheets.rake]: I expect rubocop will get grumpy at this puts?

**Review by ben-rabid — CHANGES_REQUESTED** [review]
I think the `updated_at` change is essential

**Review by ben-rabid — APPROVED** [review]

File: app/models/vocab_sheet.rb [hunk 4]
```diff
@@ -1,9 +1,14 @@
...
 
   def includes_sign?(sign_id)
     items.any? { |i| i.sign_id == sign_id }
   end
+
+  def self.clear_vocab_sheet
```
Comment (elspeth-rabid) [inline app/models/vocab_sheet.rb]: This is pretty 🐘 but I would call it `delete_old_vocab_sheets`

File: app/models/vocab_sheet.rb [hunk 5]
```diff
@@ -1,9 +1,14 @@
...
+    max_days = 31
```
Comment (elspeth-rabid) [inline app/models/vocab_sheet.rb]: Perhaps max days could be a constant at the top of the model?

MAXIMUM_AGE_IN_DAYS = 31

File: app/models/vocab_sheet.rb (hunk 4 above)
Comment (elspeth-rabid) [reply app/models/vocab_sheet.rb]: The pluralisation is the main thing, `clear_vocab_sheets` is good too

File: spec/models/vocab_sheet_spec.rb [hunk 6]
```diff
@@ -0,0 +1,16 @@
+require 'rails_helper'
+
+RSpec.describe 'VocabSheet', type: :model do
+  describe '#clear_vocab_sheet' do
+    it 'only deletes a vocab sheet older than 31 days' do
+      FactoryBot.create_list(:vocab_sheet, 5)
+      vocab_sheet = FactoryBot.create(:vocab_sheet, created_at: 1.year.ago, updated_at: 1.year.ago)
```
Comment (elspeth-rabid) [inline spec/models/vocab_sheet_spec.rb]: I would make the old record as close to 31 days as possible to test the edge case - like: `31.days.ago - 1.minute` or `32.days.ago`

File: spec/models/vocab_sheet_spec.rb [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+      FactoryBot.create_list(:item, 3, vocab_sheet_id: vocab_sheet.id)
+
```
Comment (elspeth-rabid) [inline spec/models/vocab_sheet_spec.rb]: It seems you are testing 2 things here, so it would be great to separate them into 2 tests.

1. `clear_vocab_sheet` deletes the sheet that is more than 31 days old and not the others - checking both the number of records and that the remaining ones have `updated_at` less than 31 days ago.

2. That when the vocab sheet is deleted, the items are too.

This is fine, it is just not super obvious what you are checkin and why based on the `it` block statement alone.

File: spec/rails_helper.rb [hunk 8]
```diff
@@ -51,4 +51,15 @@
   # The different available types are documented in the features, such as in
   # https://relishapp.com/rspec/rspec-rails/docs
   config.infer_spec_type_from_file_location!
+
```
Comment (elspeth-rabid) [inline spec/rails_helper.rb]: 👍

**Review by elspeth-rabid — APPROVED** [review]
Nice! I made a few comments on naming and testing stuff but nothing to hold up merging.

File: spec/models/vocab_sheet_spec.rb (hunk 7 above)
Comment (Br3nda) [reply spec/models/vocab_sheet_spec.rb]: i love that there are tests now.... so we can be more confident future devs won't make a mistake and delete all data.

File: lib/tasks/clear_vocab_sheets.rake [hunk 9]
```diff
@@ -0,0 +1,5 @@
+namespace :vocab_sheets do
+  task clear: :environment do
+    VocabSheet.clear_vocab_sheet
```
Comment (Br3nda) [inline lib/tasks/clear_vocab_sheets.rake]: the name doesn't quite capture what it does.
maybe `purge_old_sheets!` <-- :elephant:

Comment (Br3nda) [reply lib/tasks/clear_vocab_sheets.rake]: aha.. Elspeth said the same thing :-D

File: app/models/vocab_sheet.rb (hunk 5 above)
Comment (Br3nda) [reply app/models/vocab_sheet.rb]: this could be in config. Then we have control of it without changing code and doing a release. We could even turn it off.

**Review by Br3nda — APPROVED** [review]
2x :elephant: comments.

**Comment by Br3nda** [comment]
So, is the database cleaner addition to make it run faster?

**Comment by KatieWright26** [comment]
Nah, I just noticed when I ran my tests multiple times that my FactoryBot
list of VocabSheets weren't being cleared, so installed the db cleaner to
make sure all tests start fresh.
